File: flutter/lib/planner/progress_tracker.dart
import 'stage_assessor.dart';
import 'response_profiler.dart';

/// Port of fitness-planner's `progress_tracker.py`.
/// 训练日志 → stage_assessor / response_profiler 的输入。
///
/// LoggedSet     = {reps, weight_kg?, rpe?, rir?}
/// LoggedExercise= {exercise_id, planned_sets, sets: [LoggedSet]}
/// LoggedSession = {date, plan_day, session_type, planned_sets, exercises, aborted, pain_flag}
/// BodyEntry     = {date, weight_kg, waist_cm?}

const _rirTrustworthy = 3;

double? epleyE1rm(double weightKg, int reps, int? rir) {
  final r = rir ?? 0;
  if (r > _rirTrustworthy) return null;
  final eff = (reps + r) > 12 ? 12 : (reps + r);
  return double.parse((weightKg * (1 + eff / 30)).toStringAsFixed(1));
}

int _ordinal(String d) => DateTime.parse(d).difference(DateTime(1, 1, 1)).inDays;

(int, int) _isoWeek(String d) {
  final dt = DateTime.parse(d);
  // ISO week: Thursday of this week determines the year+week.
  final thursday = dt.add(Duration(days: 3 - ((dt.weekday + 6) % 7)));
  final firstThursday = DateTime(thursday.year, 1, 4);
  final week = 1 +
      (thursday.difference(firstThursday).inDays -
              ((firstThursday.weekday + 6) % 7) +
              ((thursday.weekday + 6) % 7)) ~/
          7;
  return (thursday.year, week);
}

int _completedSets(Map s) => [
      for (final ex in (s['exercises'] as List? ?? const []))
        (ex['sets'] as List? ?? const []).length,
    ].fold(0, (a, b) => a + b);

bool _isCompleted(Map s) {
  if (s['aborted'] == true) return false;
  final planned = (s['planned_sets'] as num?)?.toInt() ?? 0;
  if (planned <= 0) return _completedSets(s) > 0;
  return _completedSets(s) / planned >= 0.5;
}

List<String> _baselineIds(Map plan) => [
      for (final b in ((plan['stage_goal'] as Map?)?['baseline_lifts'] as List? ?? const []))
        b['exercise_id'] as String,
    ];

Map<String, (int, int)> _planRepsRange(Map plan) {
  final out = <String, (int, int)>{};
  for (final s in ((plan['training'] as Map)['schedule'] as List)) {
    for (final e in (s['exercises'] as List)) {
      final rr = (e['reps'] as String? ?? '').replaceAll('＋', '+');
      if (rr.contains('-')) {
        final parts = rr.split('-');
        final a = int.tryParse(parts[0]), b = int.tryParse(parts[1]);
        if (a != null && b != null) out[e['exercise_id'] as String] = (a, b);
      }
    }
  }
  return out;
}

Map<String, List<Map<String, dynamic>>> _liftSeries(
    Map plan, List<Map<String, dynamic>> sessions) {
  final repsRange = _planRepsRange(plan);
  final wanted = _baselineIds(plan).toSet();
  final series = <String, List<Map<String, dynamic>>>{for (final i in wanted) i: []};
  final ordered = [...sessions]
    ..sort((a, b) => (a['date'] as String).compareTo(b['date'] as String));
  for (final sess in ordered) {
    if (!_isCompleted(sess)) continue;
    for (final ex in (sess['exercises'] as List? ?? const [])) {
      final eid = ex['exercise_id'] as String?;
      if (eid == null || !wanted.contains(eid)) continue;
      final sets = (ex['sets'] as List? ?? const []);
      final e1rms = <double>[];
      final loads = <double>[];
      final doneReps = <int>[];
      for (final s in sets) {
        final w = (s['weight_kg'] as num?)?.toDouble();
        final reps = (s['reps'] as num?)?.toInt();
        final rir = (s['rir'] as num?)?.toInt();
        if (reps != null) doneReps.add(reps);
        if (w != null) loads.add(w);
        if (w != null && reps != null) {
          final v = epleyE1rm(w, reps, rir);
          if (v != null) e1rms.add(v);
        }
      }
      final topLoad = loads.isEmpty ? null : loads.reduce((a, b) => a > b ? a : b);
      final rr = repsRange[eid] ?? (6, 12);
      final repsAtTop = [
        for (final s in sets)
          if ((s['weight_kg'] as num?)?.toDouble() == topLoad && s['reps'] != null)
            (s['reps'] as num).toInt(),
      ];
      final rirVals = [
        for (final s in sets)
          if (s['rir'] != null) (s['rir'] as num).toInt(),
      ];
      series[eid]!.add({
        'date': sess['date'],
        'best_e1rm': e1rms.isEmpty ? null : e1rms.reduce((a, b) => a > b ? a : b),
        'top_load': topLoad,
        'best_reps_at_top': repsAtTop.isEmpty
            ? null
            : repsAtTop.reduce((a, b) => a > b ? a : b),
        'all_top_range': doneReps.isNotEmpty && doneReps.every((r) => r >= rr.$2),
        'any_below_bottom': doneReps.isNotEmpty && doneReps.any((r) => r < rr.$1),
        'avg_rir': rirVals.isEmpty
            ? null
            : double.parse(
                (rirVals.fold(0, (a, b) => a + b) / rirVals.length).toStringAsFixed(1)),
      });
    }
  }
  return series;
}

double _median(List<double> xs) {
  final s = [...xs]..sort();
  final n = s.length;
  return n.isOdd ? s[n ~/ 2] : (s[n ~/ 2 - 1] + s[n ~/ 2]) / 2;
}

(double?, int) _performancePct(Map<String, List<Map<String, dynamic>>> series) {
  final pcts = <double>[];
  var confirmations = 0;
  for (final entries in series.values) {
    final pts = [
      for (final e in entries)
        if (e['best_e1rm'] != null) e['best_e1rm'] as double,
    ];
    if (pts.length < 2) continue;
    final change = (pts.last - pts.first) / pts.first * 100;
    pcts.add(change);
    if (change >= 2.5) confirmations++;
  }
  if (pcts.isEmpty) return (null, 0);
  return (double.parse(_median(pcts).toStringAsFixed(1)), confirmations);
}

int _sameLoadRepGain(Map<String, List<Map<String, dynamic>>> series) {
  var best = 0;
  for (final entries in series.values) {
    Map<String, dynamic>? first;
    for (final e in entries) {
      if (e['top_load'] != null && e['best_reps_at_top'] != null) {
        first = e;
        break;
      }
    }
    if (first == null) continue;
    for (final e in entries.reversed) {
      if (e['top_load'] == first['top_load'] && e['best_reps_at_top'] != null) {
        final g = (e['best_reps_at_top'] as int) - (first['best_reps_at_top'] as int);
        if (g > best) best = g;
        break;
      }
    }
  }
  return best;
}

double _fitSlope(List<(double, double)> pairs) {
  final n = pairs.length;
  if (n < 2) return 0;
  final sx = pairs.fold(0.0, (a, p) => a + p.$1);
  final sy = pairs.fold(0.0, (a, p) => a + p.$2);
  final sxx = pairs.fold(0.0, (a, p) => a + p.$1 * p.$1);
  final sxy = pairs.fold(0.0, (a, p) => a + p.$1 * p.$2);
  final denom = n * sxx - sx * sx;
  return denom == 0 ? 0 : (n * sxy - sx * sy) / denom;
}

(bool, double?, double?) _bodyTrend(List<Map<String, dynamic>> body, String goal) {
  if (body.length < 3) return (false, null, null);
  final b = [...body]..sort((x, y) => (x['date'] as String).compareTo(y['date'] as String));
  final d0 = _ordinal(b.first['date'] as String);
  if (_ordinal(b.last['date'] as String) - d0 < 10) return (false, null, null);
  final w = [
    for (final e in b)
      if (e['weight_kg'] != null)
        ((_ordinal(e['date'] as String) - d0).toDouble(), (e['weight_kg'] as num).toDouble()),
  ];
  double? weightPct;
  if (w.length >= 2 && w.first.$2 != 0) {
    weightPct = double.parse(((w.last.$2 - w.first.$2) / w.first.$2 * 100).toStringAsFixed(2));
  }
  final ws = [
    for (final e in b)
      if (e['waist_cm'] != null)
        ((_ordinal(e['date'] as String) - d0).toDouble(), (e['waist_cm'] as num).toDouble()),
  ];
  double? waistPct;
  if (ws.length >= 2 && ws.first.$2 != 0) {
    waistPct = double.parse(((ws.last.$2 - ws.first.$2) / ws.first.$2 * 100).toStringAsFixed(2));
  }
  bool met;
  if (goal == 'fat_loss') {
    final wtOk = w.length >= 3 && _fitSlope(w) < 0 && (weightPct ?? 0) <= -0.5;
    final waistOk = ws.isNotEmpty && ws.first.$2 - ws.last.$2 >= 1.0;
    met = wtOk || waistOk;
  } else {
    final wtOk = w.length >= 3 && _fitSlope(w) >= 0 && (weightPct ?? 0) >= 0.25;
    final waistOk = waistPct == null || waistPct <= 1.0;
    met = wtOk && waistOk;
  }
  return (met, weightPct, waistPct);
}

/// 体重周变化率（%/周），用回归斜率算，抗单点波动。
/// 需 ≥3 次称重、跨度 ≥14 天，否则 null（数据不够不调整）。
double? weeklyWeightPct(List<Map<String, dynamic>> body) {
  final b = [
    for (final e in body)
      if (e['weight_kg'] != null && e['date'] != null) e,
  ]..sort((x, y) => (x['date'] as String).compareTo(y['date'] as String));
  if (b.length < 3) return null;
  final d0 = _ordinal(b.first['date'] as String);
  final span = _ordinal(b.last['date'] as String) - d0;
  if (span < 14) return null;
  final pairs = [
    for (final e in b)
      ((_ordinal(e['date'] as String) - d0).toDouble(),
          (e['weight_kg'] as num).toDouble()),
  ];
  final slopePerDay = _fitSlope(pairs);
  final meanW = pairs.fold(0.0, (a, p) => a + p.$2) / pairs.length;
  if (meanW <= 0) return null;
  return double.parse((slopePerDay * 7 / meanW * 100).toStringAsFixed(3));
}

bool _painUnresolved(List<Map<String, dynamic>> sessions) {
  if (sessions.isEmpty) return false;
  final ordered = [...sessions]
    ..sort((a, b) => (a['date'] as String).compareTo(b['date'] as String));
  return ordered
      .sublist(ordered.length - (ordered.length < 3 ? ordered.length : 3))
      .any((s) => s['pain_flag'] == true);
}

StageEvidence aggregateEvidence(
  Map plan,
  List<Map<String, dynamic>> workoutLog,
  List<Map<String, dynamic>> bodyLog,
) {
  final goal = (plan['profile'] as Map?)?['goal'] as String? ?? 'hypertrophy';
  final completed = workoutLog.where(_isCompleted).toList();
  final activeWeeks = {
    for (final s in completed)
      if (s['date'] != null) _isoWeek(s['date'] as String),
  }.length;
  final series = _liftSeries(plan, workoutLog);
  final comparable = series.values
      .map((v) => v.where((e) => e['best_e1rm'] != null).length)
      .fold(0, (a, b) => a > b ? a : b);
  final (perfPct, confirmations) = _performancePct(series);
  final repGain = _sameLoadRepGain(series);
  final (bodyMet, _, _) = _bodyTrend(bodyLog, goal);
  return StageEvidence(
    completedSessions: completed.length,
    activeWeeks: activeWeeks,
    comparableMeasurements: comparable,
    performanceImprovementPct: perfPct,
    performanceConfirmations: confirmations,
    sameLoadRepGain: repGain == 0 ? null : repGain,
    bodyTrendTargetMet: bodyMet,
    unresolvedSafetyIssue: _painUnresolved(workoutLog),
  );
}

ResponseObservation aggregateObservation(
  Map plan,
  List<Map<String, dynamic>> workoutLog,
  List<Map<String, dynamic>> bodyLog,
  int completedCycles,
) {
  final goal = (plan['profile'] as Map?)?['goal'] as String? ?? 'hypertrophy';
  final planned =
      ((plan['stage_goal'] as Map?)?['planned_sessions'] as num?)?.toInt() ?? 1;
  final completed = workoutLog.where(_isCompleted).length;
  final adherence = double.parse((completed / planned * 100).toStringAsFixed(1));
  final series = _liftSeries(plan, workoutLog);
  final (perfPct, _) = _performancePct(series);
  final (_, weightPct, waistPct) = _bodyTrend(bodyLog, goal);
  final rirGaps = [
    for (final v in series.values)
      for (final e in v)
        if (e['avg_rir'] != null) 2 - (e['avg_rir'] as double),
  ];
  final recovery = rirGaps.isEmpty
      ? null
      : double.parse(
          (rirGaps.fold(0.0, (a, b) => a + b) / rirGaps.length).toStringAsFixed(2));
  return ResponseObservation(
    completedCycles: completedCycles,
    adherencePct: adherence,
    performanceImprovementPct: perfPct,
    weightTrendPct: weightPct,
    waistTrendPct: waistPct,
    recoveryScore: recovery,
  );
}

int _trailingTrue(List<Map<String, dynamic>> entries, String key) {
  var n = 0;
  for (final e in entries.reversed) {
    if (e[key] == true) {
      n++;
    } else {
      break;
    }
  }
  return n;
}

Map<String, Map<String, dynamic>> perExerciseProgress(
  Map plan,
  List<Map<String, dynamic>> workoutLog,
) {
  final series = _liftSeries(plan, workoutLog);
  final out = <String, Map<String, dynamic>>{};
  series.forEach((eid, entries) {
    if (entries.isEmpty) return;
    final last3 = entries.sublist(entries.length - (entries.length < 3 ? entries.length : 3));
    final decl = entries.length >= 3 &&
        entries[entries.length - 1]['best_e1rm'] != null &&
        entries[entries.length - 2]['best_e1rm'] != null &&
        entries[entries.length - 3]['best_e1rm'] != null &&
        (entries[entries.length - 1]['best_e1rm'] as double) <
            (entries[entries.length - 2]['best_e1rm'] as double) &&
        (entries[entries.length - 2]['best_e1rm'] as double) <
            (entries[entries.length - 3]['best_e1rm'] as double);
    out[eid] = {
      'instances': entries.length,
      'e1rm_change_pct': (entries.length >= 2 &&
              entries.first['best_e1rm'] != null &&
              entries.last['best_e1rm'] != null)
          ? double.parse((((entries.last['best_e1rm'] as double) -
                      (entries.first['best_e1rm'] as double)) /
                  (entries.first['best_e1rm'] as double) *
                  100)
              .toStringAsFixed(1))
          : null,
      'last_all_top_range': last3.isNotEmpty && (last3.last['all_top_range'] == true),
      'consecutive_below_bottom': _trailingTrue(entries, 'any_below_bottom'),
      'avg_rir': last3.isNotEmpty ? last3.last['avg_rir'] : null,
      'e1rm_declining': decl,
    };
  });
  return out;
}

File: fitness-planner/engine/progress_tracker.py
"""ProgressTracker — 训练日志 → 阶段评估证据（缺的中间层）。

把 App 记录的「实际做了什么」聚合成 stage_assessor / response_profiler 要的输入。

日志契约：
    LoggedSet     = {reps, weight_kg?, rpe?, rir?}
    LoggedExercise= {exercise_id, planned_sets, sets: [LoggedSet]}
    LoggedSession = {date "YYYY-MM-DD", plan_day, session_type, planned_sets,
                     exercises: [LoggedExercise], aborted, pain_flag}
    BodyEntry     = {date "YYYY-MM-DD", weight_kg, waist_cm?}
"""
from __future__ import annotations

from datetime import date
from statistics import median

from .stage_assessor import StageEvidence
from .response_profiler import ResponseObservation


_RIR_TRUSTWORTHY = 3   # RIR > 3 的组离力竭太远，不能用来估 1RM


def epley_e1rm(weight_kg: float, reps: int, rir: int | None) -> float | None:
    """Epley + RIR 修正。RIR 缺失按 0（练到力竭）；RIR > 3 视为不可信，返回 None。"""
    r = rir if rir is not None else 0
    if r > _RIR_TRUSTWORTHY:
        return None
    eff = min(12, reps + r)
    return round(weight_kg * (1 + eff / 30), 1)


def _iso_week(d: str) -> tuple[int, int]:
    y, m, dd = (int(x) for x in d.split("-"))
    iso = date(y, m, dd).isocalendar()
    return (iso[0], iso[1])


def _completed_sets(sess: dict) -> int:
    return sum(len(ex.get("sets", [])) for ex in sess.get("exercises", []))


def _is_completed(sess: dict) -> bool:
    if sess.get("aborted"):
        return False
    planned = sess.get("planned_sets") or 0
    if planned <= 0:
        return _completed_sets(sess) > 0
    return _completed_sets(sess) / planned >= 0.5


def _baseline_ids(plan: dict) -> list[str]:
    return [b["exercise_id"] for b in plan.get("stage_goal", {}).get("baseline_lifts", [])]


def _lift_series(plan: dict, sessions: list[dict]) -> dict[str, list[dict]]:
    """每个基准动作按时间排的 [{date, best_e1rm, best_reps_at_load, top_load, all_top_range, all_below_bottom}]。"""
    reps_range = _plan_reps_range(plan)
    wanted = set(_baseline_ids(plan))
    series: dict[str, list[dict]] = {i: [] for i in wanted}
    for sess in sorted(sessions, key=lambda s: s.get("date", "")):
        if not _is_completed(sess):
            continue
        for ex in sess.get("exercises", []):
            eid = ex.get("exercise_id")
            if eid not in wanted:
                continue
            sets = ex.get("sets", [])
            e1rms = [
                epley_e1rm(s["weight_kg"], s["reps"], s.get("rir"))
                for s in sets
                if s.get("weight_kg") and s.get("reps")
            ]
            e1rms = [v for v in e1rms if v is not None]
            loads = [s["weight_kg"] for s in sets if s.get("weight_kg")]
            top_load = max(loads) if loads else None
            lo, hi = reps_range.get(eid, (6, 12))
            done_reps = [s["reps"] for s in sets if s.get("reps") is not None]
            entry = {
                "date": sess.get("date"),
                "best_e1rm": max(e1rms) if e1rms else None,
                "top_load": top_load,
                "best_reps_at_top": max(
                    (s["reps"] for s in sets if s.get("weight_kg") == top_load and s.get("reps")),
                    default=None,
                ),
                "all_top_range": bool(done_reps) and all(r >= hi for r in done_reps),
                "any_below_bottom": bool(done_reps) and any(r < lo for r in done_reps),
                "avg_rir": (
                    round(sum(s["rir"] for s in sets if s.get("rir") is not None)
                          / max(1, sum(1 for s in sets if s.get("rir") is not None)), 1)
                    if any(s.get("rir") is not None for s in sets) else None
                ),
            }
            series[eid].append(entry)
    return series


def _plan_reps_range(plan: dict) -> dict[str, tuple[int, int]]:
    out: dict[str, tuple[int, int]] = {}
    for s in plan.get("training", {}).get("schedule", []):
        for e in s.get("exercises", []):
            rr = str(e.get("reps", "")).replace("＋", "+")
            if "-" in rr:
                a, b = rr.split("-", 1)
                try:
                    out[e["exercise_id"]] = (int(a), int(b))
                except ValueError:
                    pass
    return out


def _performance_pct(series: dict[str, list[dict]]) -> tuple[float | None, int]:
    """各基准动作 e1RM 首末变化的中位数 %，以及有 ≥2.5% 提升的动作数。"""
    pcts: list[float] = []
    confirmations = 0
    for entries in series.values():
        pts = [e["best_e1rm"] for e in entries if e["best_e1rm"] is not None]
        if len(pts) < 2:
            continue
        change = (pts[-1] - pts[0]) / pts[0] * 100
        pcts.append(change)
        if change >= 2.5:
            confirmations += 1
    if not pcts:
        return None, 0
    return round(median(pcts), 1), confirmations


def _same_load_rep_gain(series: dict[str, list[dict]]) -> int:
    best = 0
    for entries in series.values():
        first = next((e for e in entries if e["top_load"] and e["best_reps_at_top"]), None)
        if not first:
            continue
        for e in reversed(entries):
            if e["top_load"] == first["top_load"] and e["best_reps_at_top"]:
                best = max(best, e["best_reps_at_top"] - first["best_reps_at_top"])
                break
    return best


def _fit_slope(pairs: list[tuple[float, float]]) -> float:
    """简单线性回归斜率（x=天序，y=值）。"""
    n = len(pairs)
    if n < 2:
        return 0.0
    sx = sum(p[0] for p in pairs); sy = sum(p[1] for p in pairs)
    sxx = sum(p[0] ** 2 for p in pairs); sxy = sum(p[0] * p[1] for p in pairs)
    denom = n * sxx - sx * sx
    return (n * sxy - sx * sy) / denom if denom else 0.0


def _body_trend(body: list[dict], goal: str) -> tuple[bool, float | None, float | None]:
    """(body_trend_target_met, weight_trend_pct, waist_trend_pct)。"""
    if len(body) < 3:
        return False, None, None
    b = sorted(body, key=lambda e: e.get("date", ""))
    d0 = _ordinal(b[0]["date"])
    span_days = _ordinal(b[-1]["date"]) - d0
    if span_days < 10:
        return False, None, None
    w = [(_ordinal(e["date"]) - d0, e["weight_kg"]) for e in b if e.get("weight_kg")]
    weight_pct = None
    if len(w) >= 2 and w[0][1]:
        weight_pct = round((w[-1][1] - w[0][1]) / w[0][1] * 100, 2)
    ws = [(_ordinal(e["date"]) - d0, e["waist_cm"]) for e in b if e.get("waist_cm")]
    waist_pct = None
    if len(ws) >= 2 and ws[0][1]:
        waist_pct = round((ws[-1][1] - ws[0][1]) / ws[0][1] * 100, 2)

    met = False
    if goal == "fat_loss":
        wt_ok = len(w) >= 3 and _fit_slope(w) < 0 and (weight_pct or 0) <= -0.5
        waist_ok = ws and ws[-1][1] is not None and ws[0][1] - ws[-1][1] >= 1.0
        met = bool(wt_ok or waist_ok)
    else:  # hypertrophy / recomposition / strength
        wt_ok = len(w) >= 3 and _fit_slope(w) >= 0 and (weight_pct or 0) >= 0.25
        waist_ok = waist_pct is not None and waist_pct <= 1.0
        met = bool(wt_ok and (waist_ok or waist_pct is None))
    return met, weight_pct, waist_pct


def _ordinal(d: str) -> int:
    y, m, dd = (int(x) for x in d.split("-"))
    return date(y, m, dd).toordinal()


def _pain_unresolved(sessions: list[dict]) -> bool:
    if not sessions:
        return False
    recent = sorted(sessions, key=lambda s: s.get("date", ""))[-3:]
    return any(s.get("pain_flag") for s in recent)


def aggregate_evidence(
    plan: dict, workout_log: list[dict], body_log: list[dict],
) -> StageEvidence:
    goal = plan.get("profile", {}).get("goal", "hypertrophy")
    completed = [s for s in workout_log if _is_completed(s)]
    active_weeks = len({_iso_week(s["date"]) for s in completed if s.get("date")})
    series = _lift_series(plan, workout_log)
    comparable = max((len([e for e in v if e["best_e1rm"] is not None]) for v in series.values()), default=0)
    perf_pct, confirmations = _performance_pct(series)
    rep_gain = _same_load_rep_gain(series)
    body_met, _, _ = _body_trend(body_log, goal)
    return StageEvidence(
        completed_sessions=len(completed),
        active_weeks=active_weeks,
        comparable_measurements=comparable,
        performance_improvement_pct=perf_pct,
        performance_confirmations=confirmations,
        same_load_rep_gain=rep_gain or None,
        body_trend_target_met=body_met,
        stable_positive_response=False,
        unresolved_safety_issue=_pain_unresolved(workout_log),
    )


def aggregate_observation(
    plan: dict, workout_log: list[dict], body_log: list[dict], completed_cycles: int,
) -> ResponseObservation:
    goal = plan.get("profile", {}).get("goal", "hypertrophy")
    planned = plan.get("stage_goal", {}).get("planned_sessions") or 1
    completed = [s for s in workout_log if _is_completed(s)]
    adherence = round(len(completed) / planned * 100, 1)
    series = _lift_series(plan, workout_log)
    perf_pct, _ = _performance_pct(series)
    _, weight_pct, waist_pct = _body_trend(body_log, goal)
    # 恢复分：实际 RIR 相对目标 RIR 2 的差（正 = 留力更多 = 更没压榨/更轻松）
    rir_gaps = [2 - e["avg_rir"] for v in series.values() for e in v if e["avg_rir"] is not None]
    recovery = round(sum(rir_gaps) / len(rir_gaps), 2) if rir_gaps else None
    return ResponseObservation(
        completed_cycles=completed_cycles,
        adherence_pct=adherence,
        performance_improvement_pct=perf_pct,
        weight_trend_pct=weight_pct,
        waist_trend_pct=waist_pct,
        recovery_score=recovery,
    )


def per_exercise_progress(plan: dict, workout_log: list[dict]) -> dict[str, dict]:
    """基准动作的进阶信号（给 check_in 做双进阶用）。"""
    series = _lift_series(plan, workout_log)
    out: dict[str, dict] = {}
    for eid, entries in series.items():
        if not entries:
            continue
        last3 = entries[-3:]
        out[eid] = {
            "instances": len(entries),
            "e1rm_change_pct": (
                round((entries[-1]["best_e1rm"] - entries[0]["best_e1rm"])
                      / entries[0]["best_e1rm"] * 100, 1)
                if len(entries) >= 2 and entries[0]["best_e1rm"] and entries[-1]["best_e1rm"]
                else None
            ),
            "last_all_top_range": bool(last3) and all(e["all_top_range"] for e in last3[-1:]),
            "consecutive_below_bottom": _trailing_true(entries, "any_below_bottom"),
            "avg_rir": last3[-1]["avg_rir"] if last3 else None,
            "e1rm_declining": (
                len(entries) >= 3
                and entries[-1]["best_e1rm"] is not None
                and entries[-2]["best_e1rm"] is not None
                and entries[-3]["best_e1rm"] is not None
                and entries[-1]["best_e1rm"] < entries[-2]["best_e1rm"] < entries[-3]["best_e1rm"]
            ),
        }
    return out


def _trailing_true(entries: list[dict], key: str) -> int:
    n = 0
    for e in reversed(entries):
        if e.get(key):
            n += 1
        else:
            break
    return n
